### PR TITLE
8291081: Some sun/tools/jstatd/TestJstatd* tests fail with "Not a percentage\: 68.31\: expected true, was false"

### PR DESCRIPTION
**Pull request by lmesnik**
The test should use the same locales in all processes, the default language should work fine.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291081](https://bugs.openjdk.org/browse/JDK-8291081): Some sun/tools/jstatd/TestJstatd* tests fail with "Not a percentage\: 68.31\: expected true, was false"

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [2702a24a](https://git.openjdk.org/jdk/pull/9798/files/2702a24a9b2f4a3a4ef765ff186b15d91b02555a)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9798/head:pull/9798` \
`$ git checkout pull/9798`

Update a local copy of the PR: \
`$ git checkout pull/9798` \
`$ git pull https://git.openjdk.org/jdk pull/9798/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9798`

View PR using the GUI difftool: \
`$ git pr show -t 9798`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9798.diff">https://git.openjdk.org/jdk/pull/9798.diff</a>

</details>

### CODE REVIEW

**Comment by plummercj** [comment]
Can you explain why setting the locale on the test but not the subprocess is causing the failure?

**Comment by lmesnik** [comment]
> Can you explain why setting the locale on the test but not the subprocess is causing the failure?

The jstatd process prints some numbers always using "en" locale and "." as decimal separators. While reading process might have different locale  ("vi" for example) and use "," as a decimal separator. So test fails to parse such numbers.

**Comment by kevinjwalls** [comment]
I tried a few locales and could not reproduce the problem, I was expecting some kind of mismatch would trigger it, but not sure what to set where to cause this...

Something else I noticed while looking however: I looked at JstatGCUtilParser.java / verify(String... valueArray)
I see it checks values are of the right type: but after almost all checks, it does a "break",
so stops iterating values.
Only percentage checks permit checking to continue. 8-)
There are three "break" statements in verify(String... valueArray) which should be "continue".
We can handle that in a separate issue.

**Comment by lmesnik** [comment]
> I tried a few locales and could not reproduce the problem, I was expecting some kind of mismatch would trigger it, but not sure what to set and where to cause this...

It is needed to verify that the locale in java has been changed and that not G1 is used. The G1 'S0' is 0.00 which is zero for all locales and test passe because of your issue. I used additional logs to check that locale was actually updated.


Also, it is needed to see language with "," as a decimal delimiter.


> 
> Something else I noticed while looking, however: I looked at JstatGCUtilParser.java / verify(String... valueArray) I see it checks values are of the right type: but after almost all checks, it does a "break", so stops iterating values. Only percentage checks permit checking to continue. 8-) There are three "break" statements in verify(String... valueArray) which should be "continue". We can handle that in a separate issue.

Thank you! that explains a lot! I've updated the fix.

**Comment by kevinjwalls** [comment]
Thanks, great to fix the verifier now if we can.

I do think that never setting user.language is good.  Forcing jstat to run with user.language=en could be wrong, as there is a risk of mismatching the printing and the verification.

BUT I don't see why I can't reproduce the problem.. The report is on Windows, and I can sign in using a non-English language but can't make the test fail!...

But then it's not the parsing in NumberFormat that fails, it's the assert afterwards.

----------System.err:(19/1147)----------
java.lang.RuntimeException: Not a percentage: 68.31: expected true, was false

Annoyingly we print the literal String value "68.31" but not the double percentage that it parsed....

**Comment by lmesnik** [comment]
> Thanks, great to fix the verifier now if we can.
> 
> I do think that never setting user.language is good. Forcing jstat to run with user.language=en could be wrong, as there is a risk of mismatching the printing and the verification.
> 
> BUT I don't see why I can't reproduce the problem.. The report is on Windows, and I can sign in using a non-English language but can't make the test fail!...

I didn't try to change system locale. I used TEST_OPTS_JAVA_OPTIONS="-Duser.language=vi" to verify that it fails wit this message and set '-Duser.language=vi' for jstat to see that test pass if all locales are same.

You could try just to print numbers in java to check if "," actually used as decimal separator. Also the "S0" should not be zero to fail. 

> 
> But then it's not the parsing in NumberFormat that fails, it's the assert afterwards.
>
Yep, the number is parsed without exception because  the "." is a thousand separator.  So it is just ignored.

>
> Annoyingly we print the literal String value "68.31" but not the double percentage that it parsed....
Fixed to print value. Now:
java.lang.RuntimeException: Not a percentage. value: 91.10 percentage: 9110.0: expected true, was false

**Comment by kevinjwalls** [comment]
OK, I did not know that user.language=vi was the trick to make it fail.  With that I see the failure (at last! 8-) ).
(I had tried other European languages that use , as a decimal separator, and don't fail.  Different thousands separator behaviour or just values that weren't out of range?...)

You get 91.10 parsed as 9110.0 which seems odd, it's not 91 thousand... But I find comments that grouping (thousands) chars might be ignored when parsing.  So mystery is I think solved, and the test is improved! 8-)

**Review by kevinjwalls — APPROVED** [review]

File: test/jdk/sun/tools/jstatd/JstatGCUtilParser.java [hunk 1]
```diff
@@ -93,21 +93,23 @@ public static void verify(String... valueArray) throws Exception {
...
-                }
-                if (type.equals(GcStatisticsType.PERCENTAGE_OR_DASH) &&
-                        value.equals("-")) {
-                    break;
+                    continue;
                 }
                 double percentage = NumberFormat.getInstance().parse(value).doubleValue();
                 assertTrue(0 <= percentage && percentage <= 100,
-                        "Not a percentage: " + value);
+                        "Not a percentage. value: " + value + " percentage: " + percentage);
```
Comment (plummercj) [inline test/jdk/sun/tools/jstatd/JstatGCUtilParser.java:112]: This doesn't look right. If the type is `INTEGER_OR_DASH` and the value is not a "-", then we don't do the `NumberFormat` check that we previously would have done if the type was `INTEGER`. Instead we treat it as a double and assert that the value is between 0 and 100. Same issue with `DOUBLE_OR_DASH`.

**Review by plummercj — APPROVED** [review]

**Pull request closed without merging**